File: src/bin/0994-rotting-oranges.py
# @leetup=info id=994 lang=python3 slug=rotting-oranges



from typing import Optional

# @leetup=code
from collections import deque


class Solution:
    def orangesRotting(self, grid: list[list[int]]) -> int:
        R = len(grid)
        C = len(grid[0])
        q = deque()
        for r in range(R):
            for c in range(C):
                if grid[r][c] == 2:
                    q.append(((r, c), 0))

        res = 0
        visited = set()
        while q:
            ((r, c), dist) = q.popleft()
            if not (0 <= r < R and 0 <= c < C):
                continue
            if (r, c) in visited:
                continue
            visited.add((r, c))
            if grid[r][c] == 0:
                continue
            if grid[r][c] == 1:
                grid[r][c] = 2
                res = dist
            q.append(((r, c + 1), dist + 1))
            q.append(((r, c - 1), dist + 1))
            q.append(((r + 1, c), dist + 1))
            q.append(((r - 1, c), dist + 1))

        for r in range(R):
            for c in range(C):
                if grid[r][c] == 1:
                    return -1

        return res


# @leetup=code

File: src/bin/0994-rotting-oranges.rs
// @leetup=info id=994 lang=rust slug=rotting-oranges

// @leetup=code
use std::collections::*;

impl Solution {
    pub fn oranges_rotting(grid: Vec<Vec<i32>>) -> i32 {
        let r_len = grid.len();
        let c_len = grid[0].len();
        let mut q = VecDeque::new();
        let mut fresh_cnt = 0;
        for r in 0..r_len {
            for c in 0..c_len {
                if grid[r][c] == 1 {
                    fresh_cnt += 1;
                } else if grid[r][c] == 2 {
                    q.push_back(((r, c), 0));
                }
            }
        }

        let mut visited = HashSet::new();
        let mut res = 0;

        while let Some(((r, c), dist)) = q.pop_front() {
            if r < r_len && c < c_len && visited.insert((r, c)) {
                if grid[r][c] == 0 {
                    continue;
                }
                if grid[r][c] == 1 {
                    // grid[r][c] = 2;
                    fresh_cnt -= 1;
                    res = dist;
                }
                q.push_back(((r, c + 1), dist + 1));
                q.push_back(((r + 1, c), dist + 1));
                q.push_back(((r, c.wrapping_sub(1)), dist + 1));
                q.push_back(((r.wrapping_sub(1), c), dist + 1));
            }
        }

        if fresh_cnt == 0 {
            res
        } else {
            -1
        }
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::oranges_rotting;
}
